File: sach/de3358.cpp
//DE3358.CPP - Tran Huu Nam - hsgtin.vn - 2024
#include <bits/stdc++.h>
using namespace std;
vector<int> P[1001];
int a[100], n,s;  int D[100001];
int main() {
    int x,y,p,t,i,j;
    for (x=2; x<=1000; x++) {
      P[x].push_back(x);
      y=x;
      p=2;
      while (y>=p*p) {
        if (y % p == 0) {
          P[x].push_back(x/p);
          while (y%p==0) y/=p;
        }
        p++;
      }
      if (y>1) P[x].push_back(x/y);
    }
    while (cin>>n) {
      t=0;
      for (i=1; i<=n; i++) {
        cin>>a[i];
        t+=a[i];
      }
      cin>>s;
      if (t<=s) {
        cout<<t<<" "<<0<<endl;
        continue;
      }
      fill(D, D+s+1, 1000);
      D[0]=0;
      t=0;
      for (i=1; i<=n; i++) {
        y=a[i];
        for (x=t; x>=0; x--) {
          for (j=0; j<P[y].size(); j++){
            p=P[y][j];
            if (x+p<=s) D[x+p]=min(D[x+p], D[x]+(j>0));
          }
          t=min(t+y, s);
        }
      }
      while (D[s]==1000) s--;
      cout<<s<<" "<<D[s]<<endl;
    }
    return 0;
}

File: sach/de3358.py
//DE3358.PY - Tran Huu Nam - hsgtin.vn - 2024
import sys
from collections import defaultdict
P = defaultdict(list)
a = [0] * 100
n, s = 0, 0
D = [1000] * 100001
def main():
    global n, s
    for x in range(2, 1001):
        P[x].append(x)
        y = x
        p = 2
        while y >= p * p:
            if y % p == 0:
                P[x].append(x // p)
                while y % p == 0:
                    y //= p
            p += 1
        if y > 1:
            P[x].append(x // y)
    for line in sys.stdin:
        n = int(line.strip())
        t = 0
        for i in range(1, n + 1):
            a[i] = int(input().strip())
            t += a[i]
        s = int(input().strip())
        if t <= s:
            print(t, 0)
            continue
        D[0] = 0
        t = 0
        for i in range(1, n + 1):
            y = a[i]
            for x in range(t, -1, -1):
                for j in range(len(P[y])):
                    p = P[y][j]
                    if x + p <= s:
                        D[x + p] = min(D[x + p], D[x] + (j > 0))
            t = min(t + y, s)
        while D[s] == 1000:
            s -= 1
        print(s, D[s])
if __name__ == "__main__":
    main()

